src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java

```diff
@@ -1873,4 +1881,69 @@ private boolean expectedFunction(@Nullable PsiElement resolve) {
       return true;
     }
   }
+
+  public static class ErlangFunctionCallParameterResolvingBySpec<T extends PsiElement> extends PatternCondition<T> {
+    public ErlangFunctionCallParameterResolvingBySpec() {
+      super("functionCallParameterResolvingBySpec");
+    }
+
+    @Override
+    public boolean accepts(@NotNull T element, ProcessingContext context) {
+      ErlangQAtom argument = element instanceof ErlangQAtom ? (ErlangQAtom) element : null;
+      ErlangSpecification specification = getSpecification(argument);
+      if (argument == null || specification == null) return false;
+
+      int position = getPositionOfArgument(argument);
+      if (position == -1) return false;
+
+      ErlangFunTypeSigs signature = specification.getSignature();
+      if (signature == null) return false;
+
+      for (ErlangTypeSig typeSig : signature.getTypeSigList()) {
+        List<ErlangType> types = typeSig.getFunType().getFunTypeArguments().getTypeList();
+        if (types.size() <= position) continue;
+
+        ErlangQVar typeVar = types.get(position).getQVar();
+        String typeVarName = typeVar != null ? typeVar.getName() : null;
+        ErlangTypeSigGuard sigGuard = typeSig.getTypeSigGuard();
+        if (sigGuard == null || typeVarName == null) continue;
+
+        for (ErlangTypeGuard guard : sigGuard.getTypeGuardList()) {
+          ErlangTopType topType = PsiTreeUtil.getChildOfType(guard, ErlangTopType.class);
+          ErlangQVar var = PsiTreeUtil.getChildOfType(topType, ErlangQVar.class);
+          if (var != null && typeVarName.equals(var.getName())) {
+            for (ErlangType typeAndVar : guard.getTypeList()) {
+              ErlangType type = typeAndVar.getType();
+              if (type != null && "module()".equals(type.getText())) {
+                return true;
+              }
+            }
+          }
+        }
+      }
```

Comment:
Please, decompose your logic. Extract a method which would calculate function argument type, and use it. Use org.intellij.erlang.types.ErlangExpressionType, and extend it's functionality.